### PR TITLE
Merge the disaggregation prototyping code

### PR DESCRIPTION
**Pull request by zhihaoshan-google**
1. disaggregation prototype
2. separate sharding for prefill and decode attention layer. (2d)
3. fix some bugs
4. add a flag to replicate the quantization scale to avoid inefficient xla fusion.

### CODE REVIEW

**Comment by vipannalla** [comment]
Looks mostly good, added few comments. This is a longish PR with changes to attention code. As chatted offline, please run couple model benchmarks and verify correctness just to be safe.

File: MaxText/layers/attentions.py [hunk 1]
```diff
@@ -577,12 +580,18 @@ def _get_cache_scale_logical_shape(self, batch, heads):
...
     dtype = self._get_cached_kv_dtype(self.dtype)
     cache_logical_shape = (batch, self.max_prefill_predict_length, heads, kv_head_size)
 
-    cache_axis_names = self.transpose_tuple(self.cache_logical_axis_names, self.prefill_cache_axis_order)
+    if model_mode == common_types.MODEL_MODE_PREFILL:
+      cache_logical_axis_names = self.prefill_cache_logical_axis_names
+    else:
+      cache_logical_axis_names = self.cache_logical_axis_names
+
+    cache_axis_names = self.transpose_tuple(cache_logical_axis_names, self.prefill_cache_axis_order)
```
[the remark this answers is not in the document]
Comment (zhihaoshan-google) [reply MaxText/layers/attentions.py:594]: It doesn't need a different layout, It's just a different axis name.

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -634,13 +648,18 @@ def _get_prefill_cache_vars(self, batch, heads, kv_head_size):
...
     cache_length = self.max_target_length - self.max_prefill_predict_length
     cache_logical_shape = (batch, cache_length, heads, kv_head_size)
 
-    cache_axis_names = self.transpose_tuple(self.cache_logical_axis_names, self.ar_cache_axis_order)
+    if model_mode == common_types.MODEL_MODE_PREFILL:
+      cache_logical_axis_names = self.prefill_cache_logical_axis_names
+    else:
+      cache_logical_axis_names = self.cache_logical_axis_names
+
+    cache_axis_names = self.transpose_tuple(cache_logical_axis_names, self.ar_cache_axis_order)
```
[the remark this answers is not in the document]
Comment (zhihaoshan-google) [reply MaxText/layers/attentions.py:662]: ar cache is also initialized in the prefill. (even though it's dummy data)

**Review by vipannalla — APPROVED** [review]
LGTM, please be aware of the quantization PR that Mitali has in flight -- https://github.com/AI-Hypercomputer/maxtext/pull/1051

File: MaxText/maxengine.py [hunk 3]
```diff
@@ -430,16 +442,15 @@ def init_decode_state(
...
           dtype=jnp.int32,
       )
       _, cache = self.model.apply(
           abstract_params,
           x,
           x,
-          decoder_segment_ids=jnp.zeros(x.shape, dtype=jnp.int32) + common_types.DECODING_ACTIVE_SEQUENCE_INDICATOR,
           enable_dropout=False,
-          model_mode=common_types.MODEL_MODE_PREFILL,
+          model_mode=common_types.MODEL_MODE_AUTOREGRESSIVE,
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/maxengine.py:456]: Please test quantization scripts for this change. Are the aqt tensors being generated correctly. You can call these scripts to verify: end_to_end/tpu/test_decode_save_quantized_ckpt.sh followed by end_to_end/tpu/test_decode_load_quantized_ckpt.sh

File: MaxText/layers/quantizations.py [hunk 4]
```diff
@@ -83,7 +83,15 @@ def _rhs_axis_metadata_wrapper(
     no_sharding_axis: Sequence[int],
     mesh_axes: Tuple[str, ...],
     is_tiled: bool,
+    replicate_scale: bool = False,
 ):
+  if replicate_scale:
```
Comment (singh-mitali) [inline MaxText/layers/quantizations.py:88]: Please add more comments on why this flag is needed and when it should be used.

Comment (singh-mitali) [reply MaxText/layers/quantizations.py:88]: As discussed - suggest breaking this into separate PRs, sharding changes, quantization and config changes. Quantization change should also be reviewed by aqt team, as there may be a better way to replicate scale supported in code.
